tests/std/tests/P0019R8_atomic_ref/test.cpp

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <atomic>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <execution>
+#include <vector>
+
+struct bigint {
+    int value;
+    int more_value[10];
+
+    bigint(int value = 0) : value(value) {}
+
+    operator int() const {
+        return value;
+    }
+};
+
+struct int128 {
+    long long value;
+    long long more_value;
+
+    int128(int value = 0) : value(value), more_value(0) {}
+
+    operator int() const {
+        return static_cast<int>(value);
+    }
+};
+
+
+// code reuse of ../P1135R6_atomic_flag_test/test.cpp
+
+template <bool AddViaCas, typename ValueType>
+void test_ops() {
+    constexpr std::size_t unique      = 80; // small to avoid overflow even for char
+    constexpr std::size_t repetitions = 8000;
+    constexpr std::size_t total       = unique * repetitions;
+    constexpr std::size_t range       = 10;
+
+    struct alignas(std::atomic_ref<ValueType>::required_alignment) Padded {
+        ValueType vals[unique] = {};
+    } padded;
+
+    auto& vals = padded.vals;
+
+    std::vector<std::atomic_ref<ValueType>> refs;
+    refs.reserve(total);
+    for (std::size_t i = 0; i != repetitions; ++i) {
+        for (auto& val : vals) {
+            refs.push_back(std::atomic_ref<ValueType>(val));
+        }
+    }
+
+    using std::execution::par;
+
+    auto load  = [](const std::atomic_ref<ValueType>& ref) { return static_cast<int>(ref.load()); };
+    auto xchg0 = [](std::atomic_ref<ValueType>& ref) { return static_cast<int>(ref.exchange(0)); };
+
+    int (*inc)(std::atomic_ref<ValueType> & ref);
+    if constexpr (AddViaCas) {
+        inc = [](std::atomic_ref<ValueType>& ref) {
+            for (;;) {
+                ValueType e = ref.load();
+                ValueType d = static_cast<ValueType>(static_cast<int>(e) + 1);
+                if (ref.compare_exchange_weak(e, d)) {
+                    return static_cast<int>(e);
+                }
+            }
+        };
+    } else {
+        inc = [](std::atomic_ref<ValueType>& ref) { return static_cast<int>(ref.fetch_add(1)); };
+    }
+
+    assert(std::transform_reduce(par, refs.begin(), refs.end(), 0, std::plus{}, load) == 0);
+    assert(std::transform_reduce(par, refs.begin(), refs.begin() + range, 0, std::plus{}, inc) == 0);
+    assert(std::transform_reduce(par, refs.begin(), refs.end(), 0, std::plus{}, load) == range * repetitions);
+    assert(std::transform_reduce(par, refs.begin(), refs.begin() + range, 0, std::plus{}, inc) == range);
+    assert(std::transform_reduce(par, refs.begin(), refs.end(), 0, std::plus{}, load) == range * repetitions * 2);
+    assert(std::transform_reduce(par, refs.begin(), refs.end(), 0, std::plus{}, xchg0) == range * 2);
+    assert(std::transform_reduce(par, refs.begin(), refs.end(), 0, std::plus{}, load) == 0);
+}
+
+#include <intrin.h>
+
+int main() {

```

Comment:
Do we have sufficient test coverage for all member functions of `atomic_ref`, including the required specializations for integral types? This isn't even verifying that all of the member functions can be instantiated.

Reply:
I've added tests for floating, integer, and pointer to try `fetch_add` / `fetch_sub`, this also covers constness of `atomic_ref` methods